File: app/src/main/java/me/dev4vin/samplepromiseapp/common/DataRepository.java
package me.dev4vin.samplepromiseapp.common;

import me.dev4vin.samplepromiseapp.data.db.AppDatabase;
import me.dev4vin.samplepromiseapp.data.net.ServerAPI;
import me.dev4vin.samplepromiseapp.error.ServerError;
import me.dev4vin.samplepromiseapp.models.Todo;
import me.yoctopus.model.List;
import me.yoctopus.model.ResponseCallBack;

public class DataRepository {
  private static DataRepository instance;

  private final AppDatabase appDatabase;
  private final ServerAPI serverAPI;

  private DataRepository() {
    appDatabase = AppDatabase.instance();
    serverAPI = ServerAPI.instance();
  }

  /**
   * get todos from the either the database or the upstream server
   * if todos are in the database, return the ones in the database hence fetch from server, store
   * in the database and return a fetch from the database
   *
   * @param skip             the todos to skip in the database
   * @param limit            the number of todos needed
   * @param responseCallBack return response back to caller
   */
  public void getTodos(int skip, int limit, final ResponseCallBack<List<Todo>, Exception> responseCallBack) {
    List<Todo> todos;
    synchronized (appDatabase) {
      todos = appDatabase.todos(skip, limit);
    }
    if (todos.isEmpty()) {
      serverAPI.getTodos(skip, limit, new ResponseCallBack<List<Todo>, ServerError>()
          .response(new ResponseCallBack.Response<List<Todo>, ServerError>() {
            @Override
            public void onResponse(List<Todo> todos) {
              synchronized (appDatabase) {
                appDatabase.saveTodos(todos);
              }
              responseCallBack.response(todos);
            }
          }).error(new ResponseCallBack.Error<ServerError>() {
            @Override
            public void onError(ServerError serverError) {
              responseCallBack.error(serverError);
            }
          }));
    } else responseCallBack.response(todos);
  }

  public static DataRepository instance() {
    if (instance == null) instance = new DataRepository();
    return instance;
  }
}

File: app/src/main/java/promise/app/common/DataRepository.kt
package promise.app.common

import promise.app.data.db.AppDatabase
import promise.app.data.db.async.AsyncAppDatabase
import promise.app.data.net.ServerAPI
import promise.app.data.net.TodoApi
import promise.app.error.AppError
import promise.app.error.ServerError
import promise.app.models.Todo
import promise.model.List
import promise.model.ResponseCallBack

class DataRepository private constructor() {

  private val appDatabase: AppDatabase = AppDatabase.instance()
  /*private val serverAPI: ServerAPI = ServerAPI.instance()*/

  /**
   * get todos from the either the database or the upstream server
   * if todos are in the database, return the ones in the database hence fetch from server, store
   * in the database and return a fetch from the database
   *
   * @param skip             the todos to skip in the database
   * @param limit            the number of todos needed
   * @param responseCallBack return response back to caller
   */
  fun getTodos(skip: Int, limit: Int, responseCallBack: ResponseCallBack<List<Todo>, Exception>) {
    val todos = appDatabase.todos(skip, limit)
    if (todos.isEmpty())
      promise.app.data.net.getTodos(ResponseCallBack<List<Todo>, AppError>()
          .response {
            appDatabase.saveTodos(it)
            responseCallBack.response(it)
          }
          .error { responseCallBack.error(it) })
    else
      responseCallBack.response(todos)

  }

  companion object {
    private var instance: DataRepository? = null

    fun instance(): DataRepository {
      if (instance == null) instance = DataRepository()
      return instance as DataRepository
    }
  }
}
